msticpy/context/tiproviders/mblookup.py

```diff
@@ -0,0 +1,317 @@
+# license information.
+# Author: Thomas Roccia - @fr0gger_
+# --------------------------------------------------------------------------
+"""
+MalwareBazaar Provider.
+
+Input can be a hash, tag, signature, filetype, clamav signature, imphash, dhash,
+tlsh, telfhash, certificate issuer info, certificate subject info, or certificate.
+
+The class can be used to:
+    - Get data about a sample from MalwareBazaar
+    - Download malware sample from MalwareBazaar
+    - Query Code Signing Certificate Blocklist (CSCB)
+    - Get the recent addition from MalwareBazaar
+
+"""
+from enum import Enum
+from typing import Set
+
+import json
+import requests
+import pandas as pd
+
+from pandas import json_normalize
+from ..._version import VERSION
+from ...common.provider_settings import get_provider_settings
+
+__version__ = VERSION
+__author__ = "Thomas Roccia | @fr0gger_"
+
+_BASE_URL = "https://mb-api.abuse.ch/api/v1/"
+
+class MBEntityType(Enum):
+    """MBEntityType: Enum class for MalwareBazaar entity types."""
+
+    HASH = "hash"
+    TAG = "tag"
+    SIGNATURE = "signature"
+    FILETYPE = "filetype"
+    CLAMAV = "clamav"
+    IMPHASH = "imphash"
+    DHASH = "dhash"
+    YARA = "yara"
+    TLSH = "tlsh"
+    TELFHASH = "telfhash"
+    CODESIGNISSUER = "issuerinfo"
+    CODESIGNSUBJECT = "subjectinfo"
+    CODESIGNSN = "certificate"
+    GIMPHASH = "gimphash"
+
+class MBlookup:
+    """ MBlookup Python Class wrapper for MalwareBazaar API.
+
+        Classes:
+            MBLookup: Class for MalwareBazaar API.
+
+        Functions:
+            lookup_ioc(self, ioc_type: MBEntityType, ioc_value: str) -> pandas.DataFrame
+            download_sample(self, sha2: str) -> str
+            get_recent(self, selector: str) -> pandas.DataFrame
+            get_cscb(self) -> pandas.DataFrame
+    """
+
+    _SUPPORTED_MB_TYPES: Set[MBEntityType] = {
+        MBEntityType.HASH,
+        MBEntityType.TAG,
+        MBEntityType.SIGNATURE,
+        MBEntityType.FILETYPE,
+        MBEntityType.CLAMAV,
+        MBEntityType.IMPHASH,
+        MBEntityType.DHASH,
+        MBEntityType.YARA,
+        MBEntityType.TLSH,
+        MBEntityType.TELFHASH,
+        MBEntityType.CODESIGNISSUER,
+        MBEntityType.CODESIGNSUBJECT,
+        MBEntityType.CODESIGNSN,
+        MBEntityType.GIMPHASH,
+    }
+
+    def __init__(self, mb_key=None):
+        """Init function to get the API key if necessary"""
+        self.mb_key = mb_key # or_get_mb_api_key()
+
+    def lookup_ioc(self, observable: str, mb_type: str, limit=10) -> pd.DataFrame:
+        """
+        Lookup an IOC in MalwareBazaar.
+        """
+
+        if MBEntityType(mb_type) not in self._SUPPORTED_MB_TYPES:
+            raise KeyError(
+                f"Property type {mb_type} not supported",
+                "Valid types are",
+                ", ".join(x.value for x in MBEntityType.__members__.values()),
+            )
+
+        if mb_type == "hash":

```

Comment:
There is a lot of very similar code in this if chain.
I wonder if we could simplify it with a generic function that gets passed a data parameter depending on the type of IoC.
Something like:


```
def _make_mb_request(data):
    res = httpx.post(_BASE_URL, data=data)
    res_data = json.loads(res.text) # or if it works ok res.json()
    if res["query_status"] == 'ok':
        return json_normalize(res['data'])
    return res["query_status"]

if mb_type == "hash":
    return _make_mb_request({'query': 'get_info', "hash": observable})
```

Reply:
Sounds good I can do that